### ISSUE TITLE
Support creation of multiple openapi docuements

### ISSUE THREAD

**Issue by Postremus** [quarkusio/quarkus#22355]
### Description

Smallrye OpenAPI supports creation of multiple openapi documents using profiles.

This is useful in e.g. following scenarios:
* Seperation by external and internal description
* Seperation by domain
* Seperation by api version

A profile is declared by defining an extension on the resource class or method that should be part of a profile:
```
@Extension(name = "x-smallrye-profile-external", value = "")
```

All created openapi schemas should be viewable in swagger-ui.

FTR, this was the PR for smallrye-openapi introducing profile support.
https://github.com/smallrye/smallrye-open-api/pull/967

**Comment by phillip-kruger** [comment]
@Postremus are you working on implementing this ?

**Comment by Postremus** [comment]
yep, will take a look. Might take a bit though for me to get started on this one.

**Comment by Postremus** [comment]
I am removing myself as asignee from this issue. TBH, I noticed everytime I tried to work on this that I am just not interested enough to do this in my free time. :)

(although yes, I want this feature for at-work stuff.)

**Comment by ttntrifork** [comment]
Ah, just came to here from google, as I was looking to see if Quarkus had support for this.

I hope somebody will pick up this feature request eventually.

**Comment by phillip-kruger** [comment]
I have no idea. Maybe @MikeEdgar  knows? I suspect we would need some work in the extension to scan more than once ? Or have a runtime filter that filter things out/in based on the `@extension`

**Comment by Postremus** [comment]
@phillip-kruger 
AFAIK:
* Allowing multiple configurations, one for each openapi document that should be generated.
* Expand the configuration to have an optional includeProfile and excludeProfile list.
* For each configurations, scan everything.
* Allow swagger-ui to display these additional generated documents

An `@extension` and Runtime Filter itself is not enough, since you can have multiple documents, and multiple profiles in one application.

**Comment by MikeEdgar** [comment]
Yes, we would need to scan multiple times, apply filters to each document at runtime (do filters apply to all or just one OpenAPI result?), etc.

Another important thing to work out is how the multiple OpenAPI documents are hosted. Currently there is only `/q/openapi`, but each of the documents would require its own endpoint presumably.0

This would probably require some map-based configurations to allow configs, filters, etc. to apply only to single OpenAPI documents where the key is the profile name that is already supported by Smallrye OpenAPI.

**Comment by Postremus** [comment]
I want this feature - so I guess I will just implement it myself :)

**Comment by Postremus** [comment]
@phillip-kruger maybe you can help me understand something?
For the `RouteBuildItem`, why set both the path (in newManagementRoute ) and the configKey? What is the purpose?


https://github.com/quarkusio/quarkus/blob/b792b260dff415f4f0c5a54873b7efd94bc656f2/extensions/smallrye-openapi/deployment/src/main/java/io/quarkus/smallrye/openapi/deployment/SmallRyeOpenApiProcessor.java#L301-L307



(Mostly in a working state, still need to write tests and docs..)

**Comment by phillip-kruger** [comment]
> [@phillip-kruger](https://github.com/phillip-kruger) maybe you can help me understand something? For the `RouteBuildItem`, why set both the path (in newManagementRoute ) and the configKey? What is the purpose?

I am not too sure... maybe @cescoffier can explain

**Comment by Postremus** [comment]
@phillip-kruger 
Figured it out in the end - sorry for not updating here.

It is unused right now. :)

See this comment: https://github.com/quarkusio/quarkus/pull/51760#discussion_r2665795324

**Comment by HerrDerb** [comment]
@Postremus Did this change the running order of the filters too?
I run a multitenant application. Therefor I have implemented a custom runtime OASFilter to return the correct security scheme for the specific tenant (request information). Otherwise, the swagger ui would not be usable anymore as the authentication config would not be correct.

This worked pretty good.

Since 3.31.0 I am getting the error that the vertx context is not set yet, when running the OASFilter. The context was created and set with the required properties in a vertx request filter than ran before the OASFilter. I suspect this pr, maybe I am wrong.

Is this a wanted change, or was this a side effect? How could I change the order again, or at least get request information into my OASFilter?

**Comment by Postremus** [comment]
Hey @HerrDerb 

That change was not intentional. 

Your OasFilter is annotated to run at runtime (which is the default)?
```
import org.eclipse.microprofile.openapi.OASFilter;

@OpenApiFilter(OpenApiFilter.RunStage.RUN)

public class MyFilter implements OASFilter {
}
```

I changed the logic around a bit on how the `OpenAPiDocumetService` is initialized. But the runtime filters should still execute at runtime
https://github.com/quarkusio/quarkus/blob/main/extensions/smallrye-openapi/runtime/src/main/java/io/quarkus/smallrye/openapi/runtime/OpenApiDocumentService.java#L166

Could you please create a new issue / bug with a reproducer? I can start looking at this later today.

**Comment by HerrDerb** [comment]
@Postremus Thanks, that would be highly appreciated  🙂 https://github.com/quarkusio/quarkus/issues/52481

**Comment by HerrDerb** [comment]
I' ve been just using the profiles to generate specific open api definitions and was surprised that still all components, also ones of endpoints that are not part of the profile, are included. Is this on purpose? and if yes, why?

**Comment by Postremus** [comment]
@HerrDerb 
https://quarkus.io/guides/openapi-swaggerui#cleanup

We can think about configuring this by default. WDYT?
(I have this property active in all my projects anyway)

**Comment by MikeEdgar** [comment]
@Postremus note that property will be deprecated with the next smallrye-open-api release, and is replaced by the more capable `mp.openapi.extensions.smallrye.remove-unused-components` that takes an array of component types. E.g. `examples,schemas`, or `*` for all unused components.

https://github.com/smallrye/smallrye-open-api#:~:text=Remove%20unused%20components

**Comment by HerrDerb** [comment]
@Postremus Ah yes, but there is another catch in my use-case 😬 We also use to have our websocket objects in the opnapi. With `mp.openapi.extensions.smallrye.remove-unused-schemas.enable=true` (also with the new property as far as i understand) these also are dropped. 
A custom build time OSAFilter also does not work for those object as  the `mp.openapi.extensions.smallrye.remove-unused-schemas.enable` is applied after the build time filters. Maybe this needs to be changed? I could open an issue for that

**Comment by MikeEdgar** [comment]
Both of the "remove-unused-*" filters should be running any time filtering occurs. The original filter that only removed schemas runs prior to any user-supplied filters, so if your own filter(s) result in schemas being abandoned, they are currently not cleaned up afterwards. The new filter will run _after_ all others, so anything left dangling is removed.

**Comment by HerrDerb** [comment]
@MikeEdgar So what if I would want to preserve specific dangling components? 
As said, we find it very handy to include out websocket objects, which obviously are dangling. Would it be an idea for the `@Schema` annotation to add a something like a "prevent-remove-dangling" flag?

**Comment by MikeEdgar** [comment]
@HerrDerb , let me think about how we can do that. I'll get back on it.

**Comment by MikeEdgar** [comment]
@HerrDerb the linked PR will provide support to preserve specific components once smallrye-open-api 4.3 is released with the change (and updated in Quarkus).
